compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignTiles.cpp

```diff
@@ -0,0 +1,437 @@
+      AMDAIE::TileOp tileOp = coreOp.getTileOp();
+      std::optional<int64_t> column = getConstantIntValue(tileOp.getCol());
+      std::optional<int64_t> row = getConstantIntValue(tileOp.getRow());
+      if (!column || !row)
+        return coreOp.emitOpError() << "has non-constant tile location";
+      tiles.insert(std::make_pair(column.value(), row.value()));
+    }
+    if (auto subviewOp = dyn_cast<memref::SubViewOp>(userOp)) {
+      return findUsersInCoreAndAddTiles(subviewOp, logicalObjectFifo, tiles);
+    } else if (auto userLogicalObjectFifo =
+                   dyn_cast<AMDAIE::LogicalObjectFifoFromMemrefOp>(userOp)) {
+      return findUsersInCoreAndAddTiles(userLogicalObjectFifo,
+                                        logicalObjectFifo, tiles);
+    }
+  }
+  return success();
+}
+
+/// Utility to clear non-local tile assignments.
+LogicalResult clearNonLocalTiles(RewriterBase &rewriter, Operation *op) {
+  op->walk([&](AMDAIE::LogicalObjectFifoFromMemrefOp objFifo) {
+    if (objFifo.getMemorySpaceAsUInt() != 2) {
+      rewriter.setInsertionPoint(objFifo);
+      SmallVector<Value> tiles;
+      rewriter.replaceOpWithNewOp<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+          objFifo, cast<LogicalObjectFifoType>(objFifo.getOutput().getType()),
+          objFifo.getMemref(), tiles);
+    }
+  });
+  return success();
+}
+
+/// Utility to duplicate global objFifos for each strided copy-like operation
+/// user to allow global logical objectFifos to be assigned to different tile
+/// locations.
+LogicalResult duplicateGlobalObjFifos(RewriterBase &rewriter, Operation *op) {
+  op->walk([&](AMDAIE::DoublyStridedCopyOpInterface copyOp) {
+    auto source = dyn_cast_if_present<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+        copyOp.getSource().getDefiningOp());
+    auto target = dyn_cast_if_present<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+        copyOp.getTarget().getDefiningOp());
+    if (source && source.getMemorySpaceAsUInt() == 0) {
+      rewriter.setInsertionPoint(copyOp);
+      auto newSource = rewriter.create<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+          rewriter.getUnknownLoc(),
+          cast<LogicalObjectFifoType>(source.getOutput().getType()),
+          source.getMemref());
+      rewriter.replaceUsesWithIf(
+          source.getOutput(), newSource.getOutput(), [&](OpOperand &use) {
+            return use.getOwner() == copyOp.getOperation();
+          });
+    }
+    if (target && target.getMemorySpaceAsUInt() == 0) {
+      rewriter.setInsertionPoint(copyOp);
+      auto newTarget = rewriter.create<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+          rewriter.getUnknownLoc(),
+          cast<LogicalObjectFifoType>(target.getOutput().getType()),
+          target.getMemref());
+      rewriter.replaceUsesWithIf(
+          target.getOutput(), newTarget.getOutput(), [&](OpOperand &use) {
+            return use.getOwner() == copyOp.getOperation();
+          });
+    }
+  });
```

Comment:
These are almost duplicate codes as the above. Maybe extract these as a separate function.

Reply:
Done